File: src/main/java/destiny/FengShui/SanYuan/Mountain.java
/**
 * @author smallufo
 * @date 2002/9/25
 * @time 上午 12:48:35
 */
package destiny.FengShui.SanYuan;

import destiny.core.chinese.Branch;
import destiny.core.chinese.Stem;
import destiny.iching.Symbol;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;

/**
 * 山
 * TODO : JDK9 JEP301 即將支援 Enhanced Enums
 * 參考： Why shouldn't Java enum literals be able to have generic type parameters?
 * http://stackoverflow.com/q/4290878/298430
 * */
public class Mountain implements Serializable {
  private Object value;

  public final static Mountain 子 = new Mountain(Branch.子);
  public final static Mountain 癸 = new Mountain(Stem.癸);
  public final static Mountain 丑 = new Mountain(Branch.丑);
  public final static Mountain 艮 = new Mountain(Symbol.艮);
  public final static Mountain 寅 = new Mountain(Branch.寅);
  public final static Mountain 甲 = new Mountain(Stem.甲);
  public final static Mountain 卯 = new Mountain(Branch.卯);
  public final static Mountain 乙 = new Mountain(Stem.乙);
  public final static Mountain 辰 = new Mountain(Branch.辰);
  public final static Mountain 巽 = new Mountain(Symbol.巽);
  public final static Mountain 巳 = new Mountain(Branch.巳);
  public final static Mountain 丙 = new Mountain(Stem.丙);
  public final static Mountain 午 = new Mountain(Branch.午);
  public final static Mountain 丁 = new Mountain(Stem.丁);
  public final static Mountain 未 = new Mountain(Branch.未);
  public final static Mountain 坤 = new Mountain(Symbol.坤);
  public final static Mountain 申 = new Mountain(Branch.申);
  public final static Mountain 庚 = new Mountain(Stem.庚);
  public final static Mountain 酉 = new Mountain(Branch.酉);
  public final static Mountain 辛 = new Mountain(Stem.辛);
  public final static Mountain 戌 = new Mountain(Branch.戌);
  public final static Mountain 乾 = new Mountain(Symbol.乾);
  public final static Mountain 亥 = new Mountain(Branch.亥);
  public final static Mountain 壬 = new Mountain(Stem.壬);

  @NotNull
  private final static Mountain[] mountainArray;
  static {
    mountainArray = new Mountain[]
    { 子 , 癸 , 丑 , 艮 , 寅 , 甲 , 卯 , 乙 , 辰 , 巽 , 巳 , 丙 ,
      午 , 丁 , 未 , 坤 , 申 , 庚 , 酉 , 辛 , 戌 , 乾 , 亥 , 壬
    };
  }

  private Mountain(Object o)
  {
    this.value = o;
  }

  public Object getValue()
  {
    return value;
  }

  @NotNull
  public static Mountain[] getMountains()
  {
    return mountainArray;
  }

  private static int getIndex(@NotNull Mountain m) {
    int index = -1;
    for (int i = 0; i < mountainArray.length; i++) {
      if (m.equals(mountainArray[i]))
        index = i;
    }
    return index;
  }

  public Mountain getOppositeMountain() {
    return getOppositeMountain(this);
  }

  public static Mountain getOppositeMountain(@NotNull Mountain m) {
    return mountainArray[normalize(Mountain.getIndex(m) + 12)];
  }

  private static int normalize(int value)
  {
    if (value > 23 )
      return normalize( value-24);
    else if (value < 0)
      return normalize (value+24);
    else
      return value;
  }

  public String toString()
  {
    return value.toString();
  }

}

File: src/main/java/destiny/core/fengshui/Mountain.kt
/**
 * Created by smallufo on 2018-02-25.
 */
package destiny.core.fengshui

import destiny.core.chinese.Branch
import destiny.core.chinese.Stem
import destiny.iching.Symbol

sealed class SealedMnt {
  data class MntBranch(val branch: Branch) : SealedMnt()
  data class MntStem(val stem: Stem) : SealedMnt()
  data class MntSymbol(val symbol: Symbol) : SealedMnt()
}

enum class Mountain(val mnt: SealedMnt, val symbol: Symbol) {
  子(SealedMnt.MntBranch(Branch.子), Symbol.坎),
  癸(SealedMnt.MntStem(Stem.癸), Symbol.坎),

  丑(SealedMnt.MntBranch(Branch.丑), Symbol.艮),
  艮(SealedMnt.MntSymbol(Symbol.艮), Symbol.艮),
  寅(SealedMnt.MntBranch(Branch.寅), Symbol.艮),

  甲(SealedMnt.MntStem(Stem.甲), Symbol.震),
  卯(SealedMnt.MntBranch(Branch.卯), Symbol.震),
  乙(SealedMnt.MntStem(Stem.乙), Symbol.震),

  辰(SealedMnt.MntBranch(Branch.辰), Symbol.巽),
  巽(SealedMnt.MntSymbol(Symbol.巽), Symbol.巽),
  巳(SealedMnt.MntBranch(Branch.巳), Symbol.巽),

  丙(SealedMnt.MntStem(Stem.丙), Symbol.離),
  午(SealedMnt.MntBranch(Branch.午), Symbol.離),
  丁(SealedMnt.MntStem(Stem.丁), Symbol.離),

  未(SealedMnt.MntBranch(Branch.未), Symbol.坤),
  坤(SealedMnt.MntSymbol(Symbol.坤), Symbol.坤),
  申(SealedMnt.MntBranch(Branch.申), Symbol.坤),

  庚(SealedMnt.MntStem(Stem.庚), Symbol.兌),
  酉(SealedMnt.MntBranch(Branch.酉), Symbol.兌),
  辛(SealedMnt.MntStem(Stem.辛), Symbol.兌),

  戌(SealedMnt.MntBranch(Branch.戌), Symbol.乾),
  乾(SealedMnt.MntSymbol(Symbol.乾), Symbol.乾),
  亥(SealedMnt.MntBranch(Branch.亥), Symbol.乾),

  壬(SealedMnt.MntStem(Stem.壬), Symbol.坎);

  /** 0 ~ 23 */
  val index: Int by lazy {
     values().indexOf(this)
  }

  val opposite: Mountain
    get() {
      val newIndex = normalize(this.index + 12)
      return values()[newIndex]
    }

  companion object {
    fun normalize(value: Int): Int {
      return when {
        value > 23 -> normalize(value - 24)
        value < 0 -> normalize(value + 24)
        else -> value
      }
    }


  }
}
